src/openloco/windows/MessageWindow.cpp

```diff
@@ -1,26 +1,683 @@
+            {
+                if (height + 39 <= dpi->y)
+                {
+                    height += 39;
+                    continue;
+                }
+
+                if (height >= dpi->y + dpi->height)
+                {
+                    height += 39;
+                    continue;
+                }
+                auto message = messagemgr::get(i);
+                char* buffer = message->messageString;
+                auto str = const_cast<char*>(stringmgr::get_string(string_ids::buffer_2039));
+
+                const size_t bufferLength = 512;
+                strncpy(str, buffer, bufferLength);
+
+                auto stringId = string_ids::black_stringid;
+
+                if (self->row_hover == i)
+                {
+                    gfx::draw_rect(dpi, 0, height, self->width, 38, (1 << 25) | palette_index::index_30);
+                    stringId = string_ids::wcolour2_stringid;
+                }
+
+                {
+                    auto args = FormatArguments();
+                    args.push(string_ids::tiny_font_date);
+                    args.push(message->date);
+
+                    gfx::draw_string_494B3F(*dpi, 0, height, colour::black, stringId, &args);
+                }
+                {
+                    auto args = FormatArguments();
+                    args.push(string_ids::buffer_2039);
+
+                    auto width = self->widgets[widx::scrollview].width() - 14;
+                    gfx::draw_string_495224(*dpi, 0, height + 6, width, colour::black, stringId, &args);
+                    height += 39;
+                }
+            }
+        }
+
+        // 0x0042A7B9
+        static void tabReset(window* self)
+        {
+            self->min_width = minWindowSize.width;
+            self->min_height = minWindowSize.height;
+            self->max_width = maxWindowSize.width;
+            self->max_height = maxWindowSize.height;
+            self->width = minWindowSize.width;
+            self->height = minWindowSize.height;
+            self->row_hover = -1;
+        }
+
+        static void initEvents()
+        {
+            events.on_mouse_up = on_mouse_up;
+            events.on_resize = on_resize;
+            events.on_update = common::on_update;
+            events.event_08 = event_08;
+            events.event_09 = event_09;
+            events.get_scroll_size = get_scroll_size;
+            events.scroll_mouse_down = scroll_mouse_down;
+            events.scroll_mouse_over = scroll_mouse_over;
+            events.tooltip = tooltip;
+            events.prepare_draw = prepare_draw;
+            events.draw = draw;
+            events.draw_scroll = draw_scroll;
+        }
+    }
     // 0x0042A3FF
     void open()
     {
-        call(0x0042A3FF);
+        //call(0x0042A3FF);
+        auto window = WindowManager::bringToFront(WindowType::messages);
+
+        if (window != nullptr)
+        {
+            if (input::is_tool_active(_toolWindowType, _toolWindowNumber))
```

Comment:
This doesn't look correct. I think you are meant to pass the window->class and window->number